sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JDBCOptions.scala

```diff
@@ -23,12 +23,15 @@ import java.util.{Locale, Properties}
 import org.apache.commons.io.FilenameUtils
 
 import org.apache.spark.SparkFiles
+import org.apache.spark.annotation.DeveloperApi
 import org.apache.spark.internal.Logging
 import org.apache.spark.sql.catalyst.util.CaseInsensitiveMap
 
 /**
+ * ::DeveloperApi::
  * Options for the JDBC data source.
  */
+@DeveloperApi
```

Comment:
This is an internal class, so could you avoid exposing it? How about using `Map[String, String]` instead in the provider?

Reply:
Good idea, this way the API is less exposed to changes.

Reply:
I've done the initial steps but I see bad and worse tradeoffs. The API is less dependent on `JDBCOptions` changes but the provider side implementation becames more complicated with either duplicated code or duplicated instantiation. Let me explain in examples:

Option1: Re-instantiate `JDBCOptions` where duplicated instantiation worries me + the general concern down below.
```
  override def canHandle(driver: Driver, options: Map[String, String]): Boolean = {
    val jdbcOptions = new JDBCOptions(options)
    jdbcOptions.keytab == null || jdbcOptions.principal == null
  }
```

Option2: Find out the needed parameters on by own where re-inventing the wheel feeling worries me + the general concern down below.
```
  override def canHandle(driver: Driver, options: Map[String, String]): Boolean = {
    val keytab = {
      val keytabParam = options.getOrElse(JDBC_KEYTAB, null)
      if (keytabParam != null && FilenameUtils.getPath(keytabParam).isEmpty) {
        val result = SparkFiles.get(keytabParam)
        logDebug(s"Keytab path not found, assuming --files, file name used on executor: $result")
        result
      } else {
        logDebug("Keytab path found, assuming manual upload")
        keytabParam
      }
    }
    val principal = options.getOrElse(JDBC_PRINCIPAL, null)
    keytab == null || principal == null
  }
```

General concern: Both cases Spark can be good because in the first case new `JDBCOptions` instantiation, in the second case copy paste moved into a base class can fill the gap. However considering myself as a 3rd-party developer I have not much options (since `JDBCOptions` is not exposed as developer API):
* Copy and paste the code from `JDBCOptions` which may change over time
* Implement the parameter parsing on my own which may contain bugs

Considering these findings I think it's better to keep `JDBCOptions`. WDTY?

Reply:
`Option 1` looks better. But, 3rd-party developers need to use `JDBCOptions` there? Or, Could we just pass the two params only?
```
  override def canHandle(driver: Driver, keytab: String, principal: String): Boolean = {
```

Reply:
We could pass the 2 params but then we limit further implementation possibilities so I would vote on the map.
At the moment there is no need other params other than `keytab` and `principal` but later providers may need further things. It's not a strong opinion, just don't want to close later possibilities. If we agree on the way I'll do the changes.

Reply:
The example I've given in the other open question [here](https://github.com/apache/spark/pull/29024#discussion_r454212309) can show how parameters other than `keytab` and `principal` can be used. Not passing the whole map would close this possibility.

Reply:
I agree with @maropu here. `JDBCOptions` is under `execution` package and meant to be private.

Reply:
@HyukjinKwon thanks for having a look!

I agree that `JDBCOptions` mustn't be exposed. Let me change the code to show `option 1`. As said passing only `keytab: String, principal: String` is not enough because not all but some of the providers need further configurations. I've started to work on this this change (unless anybody has better option).

Reply:
I've removed `JDBCOptions` from the API but keeping this open if further discussion needed.